From c5e423139a31ba008e80b8028f3b366f25f71073 Mon Sep 17 00:00:00 2001
From: Vincent Chen <vincent.chen@sifive.com>
Date: Fri, 19 Mar 2021 15:40:43 +0800
Subject: [PATCH 42/50] board: sifive: spl: reset USB hub, PCIe-USB bridge and
 ULPI device

Reset USB hub, PCIe-USB bridge, and ULPI device to ensure these devices
will be reset in all situations.
---
 board/sifive/hifive_unmatched_fu740/spl.c | 37 +++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/board/sifive/hifive_unmatched_fu740/spl.c b/board/sifive/hifive_unmatched_fu740/spl.c
index c1a839122f..0dfef79a07 100644
--- a/board/sifive/hifive_unmatched_fu740/spl.c
+++ b/board/sifive/hifive_unmatched_fu740/spl.c
@@ -16,6 +16,9 @@
 #include <asm/arch/gpio.h>
 #include <asm/arch/spl.h>
 
+#define UBRDG_RESET	SIFIVE_GENERIC_GPIO_NR(0, 7)
+#define ULPI_RESET	SIFIVE_GENERIC_GPIO_NR(0, 9)
+#define UHUB_RESET	SIFIVE_GENERIC_GPIO_NR(0, 11)
 #define GEM_PHY_RESET	SIFIVE_GENERIC_GPIO_NR(0, 12)
 
 #define MODE_SELECT_REG		0x1000
@@ -47,6 +50,7 @@ static inline int spl_reset_device_by_gpio(const char *label, int pin, int width
 	return ret;
 }
 
+
 static inline int spl_gemgxl_init(void)
 {
 	int ret;
@@ -61,6 +65,21 @@ static inline int spl_gemgxl_init(void)
 	return ret;
 }
 
+static inline int spl_usb_pcie_bridge_init(void)
+{
+	return spl_reset_device_by_gpio("usb_pcie_bridge_reset", UBRDG_RESET, 3000);
+}
+
+static inline int spl_usb_hub_init(void)
+{
+	return spl_reset_device_by_gpio("usb_hub_reset", UHUB_RESET, 100);
+}
+
+static inline int spl_ulpi_init(void)
+{
+	return spl_reset_device_by_gpio("ulpi_reset", ULPI_RESET, 1);
+}
+
 int spl_board_init_f(void)
 {
 	int ret;
@@ -77,6 +96,24 @@ int spl_board_init_f(void)
 		goto end;
 	}
 
+	ret = spl_usb_pcie_bridge_init();
+	if (ret) {
+		debug("USB Bridge (ASM1042A) init failed: %d\n", ret);
+		goto end;
+	}
+
+	ret = spl_usb_hub_init();
+	if (ret) {
+		debug("USB Hub (ASM1074) init failed: %d\n", ret);
+		goto end;
+	}
+
+	ret = spl_ulpi_init();
+	if (ret) {
+		debug("USB 2.0 PHY (USB3320C) init failed: %d\n", ret);
+		goto end;
+	}
+
 end:
 	return ret;
 }
-- 
2.31.0

